File: sdks/go/domain/decision.go
package domain

// FlagType names the requested flag value type for typed evaluation
// (spec/control-points.md "The nine methods").
//
// Exactly four: boolean, string, number, object — there is no separate
// integer/float distinction in v1 (Decision.Value is jsonValue;
// GetNumberValue returns number, not integer —
// conformance/surface/control-points.surface.json). The pre-v1 Go surface
// exposed a five-way Boolean/String/Integer/Float/Object split; this is the
// drift the v1 cut brings in line with node/java/python.
type FlagType string

const (
	FlagTypeBoolean FlagType = "boolean"
	FlagTypeString  FlagType = "string"
	FlagTypeNumber  FlagType = "number"
	FlagTypeObject  FlagType = "object"
)

// Reason is the normalized resolution reason (spec/decision.schema.json).
type Reason string

const (
	ReasonStatic         Reason = "STATIC"
	ReasonDefault        Reason = "DEFAULT"
	ReasonTargetingMatch Reason = "TARGETING_MATCH"
	ReasonSplit          Reason = "SPLIT"
	ReasonCached         Reason = "CACHED"
	ReasonDisabled       Reason = "DISABLED"
	ReasonStale          Reason = "STALE"
	ReasonError          Reason = "ERROR"
)

// Stable flag-metadata keys exposed under the fireweave.* namespace
// (spec/decision.schema.json standardMetadataKeys).
//
// MetaPayload (task-10b item 5, contracts/evaluation/eval-payload-
// attached.json): despite this file's prior claim that "v1 carries no
// fireweave.payload key... cut alongside the extension surface", node's own
// EvaluateOptions.includePayload was never actually cut (application/
// runtime.ts attaches it unconditionally when the caller opts in) — v1 reads
// remain side-effect free (spec/control-points.md "Side effects": no
// telemetry is EMITTED as a consequence of a read), which is an orthogonal
// concern from surfacing metadata the resolved flag already carries. That
// prior claim was the divergence itself, not a description of the ratified
// surface; this key restores go to node/python parity.
const (
	MetaErrorKind    = "fireweave.errorKind"
	MetaFlagVersion  = "fireweave.flagVersion"
	MetaVendorFlagID = "fireweave.vendorFlagId"
	MetaReasonCode   = "fireweave.reasonCode"
	MetaQuotaLimited = "fireweave.quotaLimited"
	MetaFromCache    = "fireweave.fromCache"
	MetaPayload      = "fireweave.payload"
)

// Decision is the normalized outcome of a flag resolution
// (spec/decision.schema.json). On error paths Value carries the caller
// default, Reason is ERROR, and Error carries the typed Fireweave error —
// defaults are returned, never thrown (spec/control-points.md "Return
// discipline").
type Decision struct {
	FlagKey  string
	Value    any
	Variant  string
	Reason   Reason
	Error    *Error
	Metadata map[string]any
}

// ErrorDecision builds the canonical error-path decision: default value,
// ERROR reason, and fireweave.errorKind metadata (plus any extras such as
// fireweave.quotaLimited).
func ErrorDecision(flagKey string, defaultValue any, err *Error, extraMeta map[string]any) Decision {
	meta := map[string]any{MetaErrorKind: string(err.Kind)}
	for k, v := range extraMeta {
		meta[k] = v
	}
	return Decision{
		FlagKey:  flagKey,
		Value:    defaultValue,
		Reason:   ReasonError,
		Error:    err,
		Metadata: meta,
	}
}

File: sdks/java/fireweave-sdk/src/main/java/ai/fireweave/sdk/domain/Decision.java
package ai.fireweave.sdk.domain;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Immutable canonical flag decision (mirrors {@code spec/decision.schema.json}), produced by a
 * {@code BackendAdapter} / {@code FireweaveRuntime}. No vendor types appear here.
 *
 * <p>On error paths {@link #value()} is always the caller-supplied default, {@link #reason()} is
 * {@code ERROR}, and {@code flagMetadata} carries {@code fireweave.errorKind}.
 *
 * <p>flagMetadata values are scalar only: Boolean, String, or Number.
 *
 * <p>v1 scope (spec/control-points.md "Side effects"): a read is side-effect free, so this type
 * carries no {@code payload} or {@code exposure} fields — those are schema-level extension
 * points that v1 does not surface (dropped from the pre-v1 Java shape alongside the cut
 * releases/exposures/signals namespaces).
 */
public final class Decision {

    private final String flagKey;
    private final JsonValue value;
    private final String variant;
    private final String reason;
    private final FireweaveError error;
    private final Map<String, Object> flagMetadata;

    private Decision(Builder b) {
        this.flagKey = Objects.requireNonNull(b.flagKey, "flagKey");
        this.value = Objects.requireNonNull(b.value, "value");
        this.variant = b.variant;
        this.reason = Objects.requireNonNull(b.reason, "reason");
        this.error = b.error;
        this.flagMetadata = Collections.unmodifiableMap(new LinkedHashMap<>(b.flagMetadata));
    }

    public static Builder builder(String flagKey) {
        return new Builder(flagKey);
    }

    public String flagKey() {
        return flagKey;
    }

    public JsonValue value() {
        return value;
    }

    public String variant() {
        return variant;
    }

    public String reason() {
        return reason;
    }

    /** Error occurrence, or null on success. */
    public FireweaveError error() {
        return error;
    }

    /** Scalar-only flagMetadata (Boolean | String | Number values). */
    public Map<String, Object> flagMetadata() {
        return flagMetadata;
    }

    public static final class Builder {
        private final String flagKey;
        private JsonValue value;
        private String variant;
        private String reason = Reasons.UNKNOWN;
        private FireweaveError error;
        private final Map<String, Object> flagMetadata = new LinkedHashMap<>();

        private Builder(String flagKey) {
            this.flagKey = flagKey;
        }

        public Builder value(JsonValue value) {
            this.value = value;
            return this;
        }

        public Builder variant(String variant) {
            this.variant = variant;
            return this;
        }

        public Builder reason(String reason) {
            this.reason = reason;
            return this;
        }

        public Builder error(FireweaveError error) {
            this.error = error;
            return this;
        }

        /** Value must be a Boolean, String or Number (scalar-only metadata contract). */
        public Builder metadata(String key, Object value) {
            if (value != null) {
                if (!(value instanceof Boolean || value instanceof String || value instanceof Number)) {
                    throw new IllegalArgumentException("flagMetadata values must be scalar, got "
                            + value.getClass().getSimpleName());
                }
                flagMetadata.put(key, value);
            }
            return this;
        }

        public Decision build() {
            return new Decision(this);
        }
    }
}
